File: Dopamine.Controls/SpectrumAnimation.vb
Imports System.Windows.Media.Animation
Imports System.ComponentModel

Public Class SpectrumAnimation
    Inherits Control

#Region "Private"
    Private mSpectrumPanel As StackPanel
    Private mSpectrumStoryBoard As Storyboard
#End Region

#Region "Properties"
    Public Property Accent As Brush
        Get
            Return CType(GetValue(AccentProperty), Brush)
        End Get

        Set(ByVal value As Brush)
            SetValue(AccentProperty, value)
        End Set
    End Property

    Public Property IsActive As Boolean
        Get
            Return CType(GetValue(IsActiveProperty), Boolean)
        End Get

        Set(ByVal value As Boolean)
            SetValue(IsActiveProperty, value)
        End Set
    End Property

    Public Property IsDisplayed As Boolean
        Get
            Return CType(GetValue(IsDisplayedProperty), Boolean)
        End Get

        Set(ByVal value As Boolean)
            SetValue(IsDisplayedProperty, value)
        End Set
    End Property
#End Region

#Region "Dependency Properties"
    Public Shared ReadOnly AccentProperty As DependencyProperty =
                      DependencyProperty.Register("Accent",
                      GetType(Brush), GetType(SpectrumAnimation),
                      New PropertyMetadata(TryCast(New BrushConverter().ConvertFromString("#1D7DD4"), SolidColorBrush)))

    Public Shared ReadOnly IsActiveProperty As DependencyProperty =
                      DependencyProperty.Register("IsActive",
                      GetType(Boolean), GetType(SpectrumAnimation),
                      New PropertyMetadata(Nothing))

    Public Shared ReadOnly IsDisplayedProperty As DependencyProperty =
                      DependencyProperty.Register("IsDisplayed",
                      GetType(Boolean), GetType(SpectrumAnimation),
                      New PropertyMetadata(Nothing))
#End Region

#Region "Construction"
    Shared Sub New()
        'This OverrideMetadata call tells the system that this element wants to provide a style that is different than its base class.
        'This style is defined in themes\generic.xaml
        DefaultStyleKeyProperty.OverrideMetadata(GetType(SpectrumAnimation), New FrameworkPropertyMetadata(GetType(SpectrumAnimation)))
    End Sub
#End Region

#Region "Public"
    Public Overrides Sub OnApplyTemplate()
        MyBase.OnApplyTemplate()

        mSpectrumPanel = DirectCast(GetTemplateChild("SpectrumPanel"), StackPanel)
        mSpectrumStoryBoard = CType(mSpectrumPanel.TryFindResource("SpectrumStoryBoard"), Storyboard)

        Dim d1 As DependencyPropertyDescriptor = DependencyPropertyDescriptor.FromProperty(SpectrumAnimation.IsActiveProperty, GetType(SpectrumAnimation))
        d1.AddValueChanged(Me, New EventHandler(Sub() Me.ToggleAnimation()))

        Dim d2 As DependencyPropertyDescriptor = DependencyPropertyDescriptor.FromProperty(SpectrumAnimation.IsDisplayedProperty, GetType(SpectrumAnimation))
        d2.AddValueChanged(Me, New EventHandler(Sub() Me.ToggleAnimation()))

        Me.ToggleAnimation()
    End Sub
#End Region

#Region "Private"
    Private Sub ToggleAnimation()

        If Me.IsDisplayed Then
            If Me.IsActive Then
                mSpectrumStoryBoard.Begin()
            Else
                mSpectrumStoryBoard.Pause()
            End If
        Else
            mSpectrumStoryBoard.Stop()
        End If
    End Sub
#End Region
End Class

File: Dopamine.Common/Controls/SpectrumAnimation.cs
using System;
using System.ComponentModel;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Media.Animation;

namespace Dopamine.Common.Controls
{
    public class SpectrumAnimation : Control
    {
        #region Variables
        private StackPanel spectrumPanel;
        private Storyboard spectrumStoryBoard;
        private DependencyPropertyDescriptor isActiveDescriptor;
        #endregion

        #region Properties
        public Brush Accent
        {
            get { return (Brush)GetValue(AccentProperty); }

            set { SetValue(AccentProperty, value); }
        }

        public bool IsActive
        {
            get { return Convert.ToBoolean(GetValue(IsActiveProperty)); }

            set { SetValue(IsActiveProperty, value); }
        }
        #endregion

        #region "Dependency Properties"
        public static readonly DependencyProperty AccentProperty = DependencyProperty.Register("Accent", typeof(Brush), typeof(SpectrumAnimation), new PropertyMetadata(new BrushConverter().ConvertFromString("#1D7DD4") as SolidColorBrush));
        public static readonly DependencyProperty IsActiveProperty = DependencyProperty.Register("IsActive", typeof(bool), typeof(SpectrumAnimation), new PropertyMetadata(null));
        #endregion

        #region Construction
        static SpectrumAnimation()
        {
            //This OverrideMetadata call tells the system that this element wants to provide a style that is different than its base class.
            //This style is defined in themes\generic.xaml
            DefaultStyleKeyProperty.OverrideMetadata(typeof(SpectrumAnimation), new FrameworkPropertyMetadata(typeof(SpectrumAnimation)));
        }
        #endregion

        #region Overrides
        public override void OnApplyTemplate()
        {
            base.OnApplyTemplate();
            
            spectrumPanel = (StackPanel)GetTemplateChild("SpectrumPanel");
            spectrumStoryBoard = (Storyboard)spectrumPanel.TryFindResource("SpectrumStoryBoard");

            this.isActiveDescriptor = DependencyPropertyDescriptor.FromProperty(SpectrumAnimation.IsActiveProperty, typeof(SpectrumAnimation));
            this.isActiveDescriptor.AddValueChanged(this, this.ValueChangedEventHandler);

            this.Unloaded += SpectrumAnimation_Unloaded;

            this.ToggleAnimation();
        }
        #endregion

        #region Private
        private void ToggleAnimation()
        {
            if (this.IsActive)
            {
                spectrumStoryBoard.Begin();
            }
            else
            {
                spectrumStoryBoard.Stop();
            }
        }

        private void SpectrumAnimation_Unloaded(object sender, RoutedEventArgs e)
        {
            // This prevents a memory leak.
            this.isActiveDescriptor.RemoveValueChanged(this, this.ValueChangedEventHandler);
        }
        #endregion

        #region Event handlers
        private void ValueChangedEventHandler(object sender, EventArgs e)
        {
            this.ToggleAnimation();
        }
        #endregion
    }
}
